x/uniswap/internal/types/msgs.go

```diff
@@ -0,0 +1,230 @@
+
+// MsgSwap Order - struct for swapping a coin
+type MsgSwapOrder struct {
+	SwapDenom  string         `json:"swap_denom"`   // The desired denomination either to be bought or sold
+	Amount     sdk.Coins      `json:"amount"`       // The specified amount to be either bought or sold
+	Bound      sdk.Int        `json:"bound"`        // If buy order, maximum amount of coins to be sold; otherwise minimum amount of coins to be bought
+	Deadline   time.Time      `json:"deadline"`     // deadline for the transaction to still be considered valid
+	Sender     sdk.AccAddress `json:"sender"`       // address swapping coin
+	Recipient  sdk.AccAddress `json:"recipient"`    // address output coin is being sent to
+	IsBuyOrder bool           `json:"is_buy_order"` // boolean indicating whether the order should be treated as a buy or sell
+}
+
+// NewMsgSwapOrder is a constructor function for MsgSwapOrder
+func NewMsgSwapOrder(
+	swapDenom string, amt sdk.Coins, bound sdk.Int, deadline time.Time,
+	sender, recipient sdk.AccAddress, isBuyOrder bool,
+) MsgSwapOrder {
+
+	return MsgSwapOrder{
+		SwapDenom:  swapDenom,
+		Amount:     amt,
+		Bound:      bound,
+		Deadline:   deadline,
+		Sender:     sender,
+		Recipient:  recipient,
+		IsBuyOrder: isBuyOrder,
+	}
+}
+
+// Route Implements Msg
+func (msg MsgSwapOrder) Route() string { return RouterKey }
+
+// Type Implements Msg
+func (msg MsgSwapOrder) Type() string { return "swap_order" }
+
+// ValidateBasic Implements Msg.
+func (msg MsgSwapOrder) ValidateBasic() sdk.Error {
+	if strings.TrimSpace(msg.SwapDenom) == "" {
+		return ErrNoDenom(DefaultCodespace)
+	}
+	// initially only support trading 1 coin only
+	if len(msg.Amount) != 1 {
+		return sdk.ErrInvalidCoins("must provide a single coin")
+	}
+	if !msg.Amount.IsValid() {
+		return sdk.ErrInvalidCoins("coin is invalid: " + msg.Amount.String())
+	}
+	if msg.Amount[0].Denom == msg.SwapDenom {
+		return ErrEqualDenom(DefaultCodespace)
+	}
+	if !msg.Bound.IsPositive() {
+		return ErrInvalidBound(DefaultCodespace, "")
+	}
+	if msg.Deadline.IsZero() {
+		return ErrInvalidDeadline(DefaultCodespace, "MsgSwapOrder")
+	}
+	if msg.Sender.Empty() {
+		return sdk.ErrInvalidAddress("invalid sender address")
+	}
+	if msg.Recipient.Empty() {
+		return sdk.ErrInvalidAddress("invalid recipient address")
+	}
+	return nil
+}
+
+// GetSignBytes Implements Msg.
+func (msg MsgSwapOrder) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(msg))
+}
+
+// GetSigners Implements Msg.
+func (msg MsgSwapOrder) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{msg.Sender}
+}
+
+/* --------------------------------------------------------------------------- */
+// MsgAddLiquidity
+/* --------------------------------------------------------------------------- */
+
+// MsgAddLiquidity - struct for adding liquidity to an exchange
+type MsgAddLiquidity struct {
+	ExchangeDenom string         `json:"exchange_denom"` // denomination of the exchange being added to
```

Comment:
Can we just wrap `ExchangeDenom` and `MaxCoins` to one `Deposit sdk.Coin` field?